When running `./mach test-wpt -r  /css/css-syntax/charset/page-utf16-css-bomless-utf16.html --headless --product servodriver`:

<details>
<summary>Before</summary>

```
Traceback (most recent call last):
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 1093, in run_func
    self.result = True, self.func(self.protocol, self.url, self.timeout)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 1189, in do_testharness
    raw_results = protocol.testdriver.run(url, self.script_resume, test_window)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 653, in run
    test_driver_message = self.get_next_message(url, script_resume, test_window)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 697, in get_next_message
    return self._get_next_message_classic(url, script_resume)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 704, in _get_next_message_classic
    return self.parent.base.execute_script(script_resume, asynchronous=True, args=[strip_server(url)])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 79, in execute_script
    return method(script, args=args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 689, in execute_async_script
    return self.send_session_command("POST", "execute/async", body)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 555, in send_session_command
    return self.send_command(method, url, body, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 516, in send_command
    raise err
webdriver.error.JavascriptErrorException: javascript error (500): JS evaluation raised an exception
```

</details>

<details>
<summary>After</summary>

```
Traceback (most recent call last):
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 1093, in run_func
    self.result = True, self.func(self.protocol, self.url, self.timeout)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 1189, in do_testharness
    raw_results = protocol.testdriver.run(url, self.script_resume, test_window)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 653, in run
    test_driver_message = self.get_next_message(url, script_resume, test_window)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 697, in get_next_message
    return self._get_next_message_classic(url, script_resume)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 704, in _get_next_message_classic
    return self.parent.base.execute_script(script_resume, asynchronous=True, args=[strip_server(url)])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 79, in execute_script
    return method(script, args=args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 689, in execute_async_script
    return self.send_session_command("POST", "execute/async", body)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 555, in send_session_command
    return self.send_command(method, url, body, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/work/servo/tests/wpt/tests/tools/webdriver/webdriver/client.py", line 516, in send_command
    raise err
webdriver.error.JavascriptErrorException: javascript error (500): window.__wptrunner_process_next_event is not a function

Remote-end stacktrace:

webdriverPromise</<@:9:8
webdriverPromise<@:11:21
@:2:38
@:16:15
```

The Python stack trace of the remote-side WPT executor seems a bit useless, but it is added in platform-independent code. Maybe I can try to figure out how to strip it so that only useful information is included in a followup.

--- a/components/script/dom/bindings/error.rs
+++ b/components/script/dom/bindings/error.rs
@@ -4,14 +4,20 @@
 
 //! Utilities to throw exceptions from Rust bindings.
 
+use std::ffi::CString;
+use std::ptr::NonNull;
 use std::slice::from_raw_parts;
 
 #[cfg(feature = "js_backtrace")]
 use backtrace::Backtrace;
+use embedder_traits::JavaScriptErrorInfo;
+use js::conversions::jsstr_to_string;
 use js::error::{throw_range_error, throw_type_error};
 #[cfg(feature = "js_backtrace")]
 use js::jsapi::StackFormat as JSStackFormat;
-use js::jsapi::{ExceptionStackBehavior, JS_ClearPendingException, JS_IsExceptionPending};
+use js::jsapi::{
+    ExceptionStackBehavior, JS_ClearPendingException, JS_GetProperty, JS_IsExceptionPending,
+};
 use js::jsval::UndefinedValue;
 use js::rust::wrappers::{JS_ErrorFromException, JS_GetPendingException, JS_SetPendingException};
 use js::rust::{HandleObject, HandleValue, MutableHandleValue};
@@ -236,7 +242,7 @@ impl ErrorInfo {
         None
     }
 
-    fn from_value(value: HandleValue, cx: SafeJSContext) -> ErrorInfo {
+    pub(crate) fn from_value(value: HandleValue, cx: SafeJSContext) -> ErrorInfo {
         if value.is_object() {
             rooted!(in(*cx) let object = value.to_object());
             if let Some(info) = ErrorInfo::from_object(object.handle(), cx) {
@@ -268,24 +274,47 @@ pub(crate) fn report_pending_exception(
     realm: InRealm,
     can_gc: CanGc,
 ) {
+    rooted!(in(*cx) let mut value = UndefinedValue());
+    if take_pending_exception(cx, value.handle_mut()) {
+        let error_info = ErrorInfo::from_value(value.handle(), cx);
+        report_error(
+            error_info,
+            value.handle(),
+            cx,
+            dispatch_event,
+            realm,
+            can_gc,
+        );
+    }
+}
+
+fn take_pending_exception(cx: SafeJSContext, value: MutableHandleValue) -> bool {
     unsafe {
         if !JS_IsExceptionPending(*cx) {
-            return;
+            return false;
         }
     }
-    rooted!(in(*cx) let mut value = UndefinedValue());
 
     unsafe {
-        if !JS_GetPendingException(*cx, value.handle_mut()) {
+        if !JS_GetPendingException(*cx, value) {
             JS_ClearPendingException(*cx);
             error!("Uncaught exception: JS_GetPendingException failed");
-            return;
+            return false;
         }
 
         JS_ClearPendingException(*cx);
     }
-    let error_info = ErrorInfo::from_value(value.handle(), cx);
+    true
+}
 
+fn report_error(
+    error_info: ErrorInfo,
+    value: HandleValue,
+    cx: SafeJSContext,
+    dispatch_event: bool,
+    realm: InRealm,
+    can_gc: CanGc,
+) {
     error!(
         "Error at {}:{}:{} {}",
         error_info.filename, error_info.lineno, error_info.column, error_info.message
@@ -304,12 +333,72 @@ pub(crate) fn report_pending_exception(
     }
 
     if dispatch_event {
-        GlobalScope::from_safe_context(cx, realm).report_an_error(
-            error_info,
-            value.handle(),
-            can_gc,
-        );
+        GlobalScope::from_safe_context(cx, realm).report_an_error(error_info, value, can_gc);
     }
+}
+
+pub(crate) fn javascript_error_info_from_error_info(
+    cx: SafeJSContext,
+    error_info: &ErrorInfo,
+    value: HandleValue,
+    _: CanGc,
+) -> JavaScriptErrorInfo {
+    let stack = || {
+        if !value.is_object() {
+            return None;
+        }
+
+        rooted!(in(*cx) let object = value.to_object());
+        let stack_name = CString::new("stack").unwrap();
+        rooted!(in(*cx) let mut stack_value = UndefinedValue());
+        if unsafe {
+            !JS_GetProperty(
+                *cx,
+                object.handle().into(),
+                stack_name.as_ptr(),
+                stack_value.handle_mut().into(),
+            )
+        } {
+            return None;
+        }
+        if !stack_value.is_string() {
+            return None;
+        }
+        let stack_string = NonNull::new(stack_value.to_string())?;
+        Some(unsafe { jsstr_to_string(*cx, stack_string) })
+    };
+
+    JavaScriptErrorInfo {
+        message: error_info.message.clone(),
+        filename: error_info.filename.clone(),
+        line_number: error_info.lineno as u64,
+        column: error_info.column as u64,
+        stack: stack(),
+    }
+}
+
+pub(crate) fn take_and_report_pending_exception_for_api(
+    cx: SafeJSContext,
+    realm: InRealm,
+    can_gc: CanGc,
+) -> Option<JavaScriptErrorInfo> {
+    rooted!(in(*cx) let mut value = UndefinedValue());
+    if !take_pending_exception(cx, value.handle_mut()) {
+        return None;
+    }
+
+    let error_info = ErrorInfo::from_value(value.handle(), cx);
+    let return_value =
+        javascript_error_info_from_error_info(cx, &error_info, value.handle(), can_gc);
+    report_error(
+        error_info,
+        value.handle(),
+        cx,
+        true, /* dispatch_event */
+        realm,
+        can_gc,
+    );
+    Some(return_value)
 }
 
 pub(crate) trait ErrorToJsval {

--- a/components/script/dom/globalscope.rs
+++ b/components/script/dom/globalscope.rs
@@ -86,7 +86,9 @@ use crate::dom::bindings::codegen::Bindings::VoidFunctionBinding::VoidFunction;
 use crate::dom::bindings::codegen::Bindings::WindowBinding::WindowMethods;
 use crate::dom::bindings::codegen::Bindings::WorkerGlobalScopeBinding::WorkerGlobalScopeMethods;
 use crate::dom::bindings::conversions::{root_from_object, root_from_object_static};
-use crate::dom::bindings::error::{Error, ErrorInfo, report_pending_exception};
+use crate::dom::bindings::error::{
+    Error, ErrorInfo, report_pending_exception, take_and_report_pending_exception_for_api,
+};
 use crate::dom::bindings::frozenarray::CachedFrozenArray;
 use crate::dom::bindings::inheritance::Castable;
 use crate::dom::bindings::refcounted::{Trusted, TrustedPromise};
@@ -2919,8 +2921,9 @@ impl GlobalScope {
 
             if !result {
                 debug!("error evaluating Dom string");
-                report_pending_exception(cx, true, InRealm::Entered(&ar), can_gc);
-                return Err(JavaScriptEvaluationError::EvaluationFailure);
+                let error_info =
+                    take_and_report_pending_exception_for_api(cx, InRealm::Entered(&ar), can_gc);
+                return Err(JavaScriptEvaluationError::EvaluationFailure(error_info));
             }
 
             maybe_resume_unwind();

--- a/components/script/dom/window.rs
+++ b/components/script/dom/window.rs
@@ -37,8 +37,8 @@ use devtools_traits::{ScriptToDevtoolsControlMsg, TimelineMarker, TimelineMarker
 use dom_struct::dom_struct;
 use embedder_traits::user_content_manager::{UserContentManager, UserScript};
 use embedder_traits::{
-    AlertResponse, ConfirmResponse, EmbedderMsg, PromptResponse, ScriptToEmbedderChan,
-    SimpleDialog, Theme, UntrustedNodeAddress, ViewportDetails, WebDriverJSError,
+    AlertResponse, ConfirmResponse, EmbedderMsg, JavaScriptEvaluationError, PromptResponse,
+    ScriptToEmbedderChan, SimpleDialog, Theme, UntrustedNodeAddress, ViewportDetails,
     WebDriverJSResult, WebDriverLoadStatus,
 };
 use euclid::default::{Point2D as UntypedPoint2D, Rect as UntypedRect, Size2D as UntypedSize2D};
@@ -121,7 +121,9 @@ use crate::dom::bindings::codegen::Bindings::WindowBinding::{
 use crate::dom::bindings::codegen::UnionTypes::{
     RequestOrUSVString, TrustedScriptOrString, TrustedScriptOrStringOrFunction,
 };
-use crate::dom::bindings::error::{Error, ErrorResult, Fallible};
+use crate::dom::bindings::error::{
+    Error, ErrorInfo, ErrorResult, Fallible, javascript_error_info_from_error_info,
+};
 use crate::dom::bindings::inheritance::{Castable, ElementTypeId, HTMLElementTypeId, NodeTypeId};
 use crate::dom::bindings::num::Finite;
 use crate::dom::bindings::refcounted::Trusted;
@@ -1604,23 +1606,26 @@ impl WindowMethods<crate::DomTypeHolder> for Window {
         }
     }
 
-    fn WebdriverCallback(&self, cx: JSContext, val: HandleValue, realm: InRealm, can_gc: CanGc) {
-        let rv = jsval_to_webdriver(cx, &self.globalscope, val, realm, can_gc);
-        let opt_chan = self.webdriver_script_chan.borrow_mut().take();
-        if let Some(chan) = opt_chan {
-            let _ = chan.send(rv);
+    fn WebdriverCallback(&self, cx: JSContext, value: HandleValue, realm: InRealm, can_gc: CanGc) {
+        let webdriver_script_sender = self.webdriver_script_chan.borrow_mut().take();
+        if let Some(webdriver_script_sender) = webdriver_script_sender {
+            let result = jsval_to_webdriver(cx, &self.globalscope, value, realm, can_gc);
+            let _ = webdriver_script_sender.send(result);
         }
     }
 
-    fn WebdriverException(&self, cx: JSContext, val: HandleValue, realm: InRealm, can_gc: CanGc) {
-        let rv = jsval_to_webdriver(cx, &self.globalscope, val, realm, can_gc);
-        let opt_chan = self.webdriver_script_chan.borrow_mut().take();
-        if let Some(chan) = opt_chan {
-            if let Ok(rv) = rv {
-                let _ = chan.send(Err(WebDriverJSError::JSException(rv)));
-            } else {
-                let _ = chan.send(rv);
-            }
+    fn WebdriverException(&self, cx: JSContext, value: HandleValue, can_gc: CanGc) {
+        let webdriver_script_sender = self.webdriver_script_chan.borrow_mut().take();
+        if let Some(webdriver_script_sender) = webdriver_script_sender {
+            let _ =
+                webdriver_script_sender.send(Err(JavaScriptEvaluationError::EvaluationFailure(
+                    Some(javascript_error_info_from_error_info(
+                        cx,
+                        &ErrorInfo::from_value(value, cx),
+                        value,
+                        can_gc,
+                    )),
+                )));
         }
     }
 

--- a/components/script/script_thread.rs
+++ b/components/script/script_thread.rs
@@ -3857,9 +3857,7 @@ impl ScriptThread {
             return_value.handle(),
             (&realm).into(),
             can_gc,
-        )
-        .map_err(|_| JavaScriptEvaluationError::SerializationError);
-
+        );
         let _ = self.senders.pipeline_to_constellation_sender.send((
             pipeline_id,
             ScriptToConstellationMessage::FinishJavaScriptEvaluation(evaluation_id, result),

--- a/components/script/webdriver_handlers.rs
+++ b/components/script/webdriver_handlers.rs
@@ -11,7 +11,8 @@ use base::generic_channel::GenericSender;
 use base::id::{BrowsingContextId, PipelineId};
 use cookie::Cookie;
 use embedder_traits::{
-    JSValue, WebDriverFrameId, WebDriverJSError, WebDriverJSResult, WebDriverLoadStatus,
+    JSValue, JavaScriptEvaluationError, JavaScriptEvaluationResultSerializationError,
+    WebDriverFrameId, WebDriverJSResult, WebDriverLoadStatus,
 };
 use euclid::default::{Point2D, Rect, Size2D};
 use hyper_serde::Serde;
@@ -406,7 +407,9 @@ unsafe fn jsval_to_webdriver_inner(
         if let Ok(element) = root_from_object::<Element>(*object, cx) {
             // If the element is stale, return error with error code stale element reference.
             if is_stale(&element) {
-                Err(WebDriverJSError::StaleElementReference)
+                Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::StaleElementReference,
+                ))
             } else {
                 Ok(JSValue::Element(
                     element
@@ -417,7 +420,9 @@ unsafe fn jsval_to_webdriver_inner(
         } else if let Ok(shadow_root) = root_from_object::<ShadowRoot>(*object, cx) {
             // If the shadow root is detached, return error with error code detached shadow root.
             if is_detached(&shadow_root) {
-                Err(WebDriverJSError::DetachedShadowRoot)
+                Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::DetachedShadowRoot,
+                ))
             } else {
                 Ok(JSValue::ShadowRoot(
                     shadow_root
@@ -428,7 +433,9 @@ unsafe fn jsval_to_webdriver_inner(
         } else if let Ok(window) = root_from_object::<Window>(*object, cx) {
             let window_proxy = window.window_proxy();
             if window_proxy.is_browsing_context_discarded() {
-                Err(WebDriverJSError::StaleElementReference)
+                Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::StaleElementReference,
+                ))
             } else if window_proxy.browsing_context_id() == window_proxy.webview_id() {
                 Ok(JSValue::Window(window.webview_id().to_string()))
             } else {
@@ -459,7 +466,9 @@ unsafe fn jsval_to_webdriver_inner(
                     Error::JSFailed,
                     CanGc::note(),
                 );
-                Err(WebDriverJSError::JSError)
+                Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::Generic,
+                ))
             }
         } else {
             clone_an_object(
@@ -471,7 +480,9 @@ unsafe fn jsval_to_webdriver_inner(
             )
         }
     } else {
-        Err(WebDriverJSError::UnknownType)
+        Err(JavaScriptEvaluationError::SerializationError(
+            JavaScriptEvaluationResultSerializationError::UnknownType,
+        ))
     }
 }
 
@@ -487,7 +498,9 @@ unsafe fn clone_an_object(
     let hashable = val.into();
     // Step 1. If value is in `seen`, return error with error code javascript error.
     if seen.contains(&hashable) {
-        return Err(WebDriverJSError::JSError);
+        return Err(JavaScriptEvaluationError::SerializationError(
+            JavaScriptEvaluationResultSerializationError::Generic,
+        ));
     }
     // Step 2. Append value to `seen`.
     seen.insert(hashable.clone());
@@ -502,11 +515,17 @@ unsafe fn clone_an_object(
         let length = match get_property_result {
             Ok(length) => match length {
                 Some(length) => length,
-                _ => return Err(WebDriverJSError::UnknownType),
+                _ => {
+                    return Err(JavaScriptEvaluationError::SerializationError(
+                        JavaScriptEvaluationResultSerializationError::UnknownType,
+                    ));
+                },
             },
             Err(error) => {
                 throw_dom_exception(cx, global_scope, error, CanGc::note());
-                return Err(WebDriverJSError::JSError);
+                return Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::Generic,
+                ));
             },
         };
         // Step 4. For each enumerable property in value, run the following substeps:
@@ -525,7 +544,9 @@ unsafe fn clone_an_object(
                 },
                 Err(error) => {
                     throw_dom_exception(cx, global_scope, error, CanGc::note());
-                    return Err(WebDriverJSError::JSError);
+                    return Err(JavaScriptEvaluationError::SerializationError(
+                        JavaScriptEvaluationResultSerializationError::Generic,
+                    ));
                 },
             }
         }
@@ -543,7 +564,9 @@ unsafe fn clone_an_object(
             )
         };
         if !succeeded {
-            return Err(WebDriverJSError::JSError);
+            return Err(JavaScriptEvaluationError::SerializationError(
+                JavaScriptEvaluationResultSerializationError::Generic,
+            ));
         }
         for id in ids.iter() {
             rooted!(in(*cx) let id = *id);
@@ -560,7 +583,9 @@ unsafe fn clone_an_object(
                 )
             };
             if !succeeded {
-                return Err(WebDriverJSError::JSError);
+                return Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::Generic,
+                ));
             }
 
             rooted!(in(*cx) let mut property = UndefinedValue());
@@ -573,22 +598,23 @@ unsafe fn clone_an_object(
                 )
             };
             if !succeeded {
-                return Err(WebDriverJSError::JSError);
+                return Err(JavaScriptEvaluationError::SerializationError(
+                    JavaScriptEvaluationResultSerializationError::Generic,
+                ));
             }
 
             if !property.is_undefined() {
                 let name = unsafe { jsid_to_string(*cx, id.handle()) };
                 let Some(name) = name else {
-                    return Err(WebDriverJSError::JSError);
+                    return Err(JavaScriptEvaluationError::SerializationError(
+                        JavaScriptEvaluationResultSerializationError::Generic,
+                    ));
                 };
 
-                if let Ok(value) =
-                    unsafe { jsval_to_webdriver_inner(*cx, global_scope, property.handle(), seen) }
-                {
-                    result.insert(name.into(), value);
-                } else {
-                    return Err(WebDriverJSError::JSError);
-                }
+                let value = unsafe {
+                    jsval_to_webdriver_inner(*cx, global_scope, property.handle(), seen)?
+                };
+                result.insert(name.into(), value);
             }
         }
         Ok(JSValue::Object(result))
@@ -622,17 +648,16 @@ pub(crate) fn handle_execute_script(
                 can_gc,
                 None, // No known `introductionType` for JS code from WebDriver
             );
-            let result = match evaluation_result {
-                Ok(_) => jsval_to_webdriver(cx, global, rval.handle(), realm, can_gc),
-                Err(_) => Err(WebDriverJSError::JSError),
-            };
+
+            let result = evaluation_result
+                .and_then(|_| jsval_to_webdriver(cx, global, rval.handle(), realm, can_gc));
 
             reply.send(result).unwrap_or_else(|err| {
                 error!("ExecuteScript Failed to send reply: {err}");
             });
         },
         None => reply
-            .send(Err(WebDriverJSError::BrowsingContextNotFound))
+            .send(Err(JavaScriptEvaluationError::DocumentNotFound))
             .unwrap_or_else(|err| {
                 error!("ExecuteScript Failed to send reply: {err}");
             }),
@@ -653,29 +678,24 @@ pub(crate) fn handle_execute_async_script(
             rooted!(in(*cx) let mut rval = UndefinedValue());
 
             let global_scope = window.as_global_scope();
-            if global_scope
-                .evaluate_js_on_global_with_result(
-                    &eval,
-                    rval.handle_mut(),
-                    ScriptFetchOptions::default_classic_script(global_scope),
-                    global_scope.api_base_url(),
-                    can_gc,
-                    None, // No known `introductionType` for JS code from WebDriver
-                )
-                .is_err()
-            {
-                reply_sender
-                    .send(Err(WebDriverJSError::JSError))
-                    .unwrap_or_else(|err| {
-                        error!("ExecuteAsyncScript Failed to send reply: {err}");
-                    });
+            if let Err(error) = global_scope.evaluate_js_on_global_with_result(
+                &eval,
+                rval.handle_mut(),
+                ScriptFetchOptions::default_classic_script(global_scope),
+                global_scope.api_base_url(),
+                can_gc,
+                None, // No known `introductionType` for JS code from WebDriver
+            ) {
+                reply_sender.send(Err(error)).unwrap_or_else(|error| {
+                    error!("ExecuteAsyncScript Failed to send reply: {error}");
+                });
             }
         },
         None => {
             reply
-                .send(Err(WebDriverJSError::BrowsingContextNotFound))
-                .unwrap_or_else(|err| {
-                    error!("ExecuteAsyncScript Failed to send reply: {err}");
+                .send(Err(JavaScriptEvaluationError::DocumentNotFound))
+                .unwrap_or_else(|error| {
+                    error!("ExecuteAsyncScript Failed to send reply: {error}");
                 });
         },
     }

--- a/components/script_bindings/codegen/Bindings.conf
+++ b/components/script_bindings/codegen/Bindings.conf
@@ -660,7 +660,7 @@ DOMInterfaces = {
 
 'Window': {
     'canGc': ['CreateImageBitmap', 'CreateImageBitmap_', 'CookieStore', 'Fetch', 'FetchLater', 'Open', 'SetInterval', 'SetTimeout', 'Stop', 'TrustedTypes', 'WebdriverCallback', 'WebdriverException'],
-    'inRealms': ['Fetch', 'GetOpener', 'WebdriverCallback', 'WebdriverException'],
+    'inRealms': ['Fetch', 'GetOpener', 'WebdriverCallback'],
     'additionalTraits': ['crate::interfaces::WindowHelpers'],
 },
 

--- a/components/servo/tests/webview.rs
+++ b/components/servo/tests/webview.rs
@@ -104,7 +104,10 @@ fn test_evaluate_javascript_basic(servo_test: &ServoTest) -> Result<(), anyhow::
     ensure!(result == Err(JavaScriptEvaluationError::CompilationFailure));
 
     let result = evaluate_javascript(servo_test, webview.clone(), "throw new Error()");
-    ensure!(result == Err(JavaScriptEvaluationError::EvaluationFailure));
+    ensure!(matches!(
+        result,
+        Err(JavaScriptEvaluationError::EvaluationFailure(_))
+    ));
 
     Ok(())
 }

--- a/components/shared/embedder/Cargo.toml
+++ b/components/shared/embedder/Cargo.toml
@@ -37,6 +37,7 @@ serde = { workspace = true }
 servo_geometry = { path = "../../geometry" }
 servo_url = { path = "../../url" }
 strum_macros = { workspace = true }
+strum = { workspace = true }
 stylo = { workspace = true }
 stylo_traits = { workspace = true }
 url = { workspace = true }

--- a/components/shared/embedder/lib.rs
+++ b/components/shared/embedder/lib.rs
@@ -35,7 +35,7 @@ use pixels::RasterImage;
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
 use servo_geometry::{DeviceIndependentIntRect, DeviceIndependentIntSize};
 use servo_url::ServoUrl;
-use strum_macros::IntoStaticStr;
+use strum_macros::{EnumMessage, IntoStaticStr};
 use style::queries::values::PrefersColorScheme;
 use style_traits::CSSPixel;
 use url::Url;
@@ -1058,11 +1058,33 @@ pub enum JSValue {
 }
 
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct JavaScriptErrorInfo {
+    pub message: String,
+    pub filename: String,
+    pub stack: Option<String>,
+    pub line_number: u64,
+    pub column: u64,
+}
+
+/// Indicates the reason that JavaScript evaluation failed due serializing issues the
+/// result of the evaluation.
+#[derive(Clone, Debug, Deserialize, EnumMessage, PartialEq, Serialize)]
+pub enum JavaScriptEvaluationResultSerializationError {
+    Generic,
+    DetachedShadowRoot,
+    StaleElementReference,
+    UnknownType,
+}
+
+/// An error that happens when trying to evaluate JavaScript on a `WebView`.
+#[derive(Clone, Debug, Deserialize, EnumMessage, PartialEq, Serialize)]
 pub enum JavaScriptEvaluationError {
-    /// The script could not be compiled
+    /// The `Document` of frame that the script was going to execute in no longer exists.
+    DocumentNotFound,
+    /// The script could not be compiled.
     CompilationFailure,
-    /// The script could not be evaluated
-    EvaluationFailure,
+    /// The script could not be evaluated.
+    EvaluationFailure(Option<JavaScriptErrorInfo>),
     /// An internal Servo error prevented the JavaSript evaluation from completing properly.
     /// This indicates a bug in Servo.
     InternalError,
@@ -1071,7 +1093,7 @@ pub enum JavaScriptEvaluationError {
     WebViewNotReady,
     /// The script executed successfully, but Servo could not serialize the JavaScript return
     /// value into a [`JSValue`].
-    SerializationError,
+    SerializationError(JavaScriptEvaluationResultSerializationError),
 }
 
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]

--- a/components/shared/embedder/webdriver.rs
+++ b/components/shared/embedder/webdriver.rs
@@ -24,7 +24,10 @@ use style_traits::CSSPixel;
 use webdriver::error::ErrorStatus;
 use webrender_api::units::DevicePixel;
 
-use crate::{JSValue, MouseButton, MouseButtonAction, ScreenshotCaptureError, TraversalId};
+use crate::{
+    JSValue, JavaScriptEvaluationError, MouseButton, MouseButtonAction, ScreenshotCaptureError,
+    TraversalId,
+};
 
 #[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
 pub struct WebDriverInputEventId(pub usize);
@@ -253,19 +256,7 @@ pub enum WebDriverScriptCommand {
     RemoveLoadStatusSender(WebViewId),
 }
 
-#[derive(Debug, Deserialize, Serialize)]
-pub enum WebDriverJSError {
-    /// Occurs when handler received an event message for a layout channel that is not
-    /// associated with the current script thread
-    BrowsingContextNotFound,
-    DetachedShadowRoot,
-    JSException(JSValue),
-    JSError,
-    StaleElementReference,
-    UnknownType,
-}
-
-pub type WebDriverJSResult = Result<JSValue, WebDriverJSError>;
+pub type WebDriverJSResult = Result<JSValue, JavaScriptEvaluationError>;
 
 #[derive(Debug, Deserialize, Serialize)]
 pub enum WebDriverFrameId {

--- a/components/webdriver_server/lib.rs
+++ b/components/webdriver_server/lib.rs
@@ -29,8 +29,9 @@ use capabilities::ServoCapabilities;
 use cookie::{CookieBuilder, Expiration, SameSite};
 use crossbeam_channel::{Sender, after, select};
 use embedder_traits::{
-    EventLoopWaker, JSValue, MouseButton, WebDriverCommandMsg, WebDriverCommandResponse,
-    WebDriverFrameId, WebDriverInputEventId, WebDriverJSError, WebDriverJSResult,
+    EventLoopWaker, JSValue, JavaScriptEvaluationError,
+    JavaScriptEvaluationResultSerializationError, MouseButton, WebDriverCommandMsg,
+    WebDriverCommandResponse, WebDriverFrameId, WebDriverInputEventId, WebDriverJSResult,
     WebDriverLoadStatus, WebDriverScriptCommand,
 };
 use euclid::{Point2D, Rect, Size2D};
@@ -2068,7 +2069,7 @@ impl Handler {
             .map(Duration::from_millis);
         let result = wait_for_script_ipc_response_with_timeout(receiver, timeout_duration)?;
 
-        self.postprocess_js_result(result)
+        self.javascript_evaluation_result_to_webdriver_response(result)
     }
 
     fn handle_execute_async_script(
@@ -2115,10 +2116,10 @@ impl Handler {
             .map(Duration::from_millis);
         let result = wait_for_script_ipc_response_with_timeout(receiver, timeout_duration)?;
 
-        self.postprocess_js_result(result)
+        self.javascript_evaluation_result_to_webdriver_response(result)
     }
 
-    fn postprocess_js_result(
+    fn javascript_evaluation_result_to_webdriver_response(
         &self,
         result: WebDriverJSResult,
     ) -> WebDriverResult<WebDriverResponse> {
@@ -2126,30 +2127,43 @@ impl Handler {
             Ok(value) => Ok(WebDriverResponse::Generic(ValueResponse(
                 serde_json::to_value(SendableJSValue(value))?,
             ))),
-            Err(WebDriverJSError::BrowsingContextNotFound) => Err(WebDriverError::new(
-                ErrorStatus::NoSuchWindow,
-                "Pipeline id not found in browsing context",
-            )),
-            Err(WebDriverJSError::JSException(_e)) => Err(WebDriverError::new(
-                ErrorStatus::JavascriptError,
-                "JS evaluation raised an exception",
-            )),
-            Err(WebDriverJSError::JSError) => Err(WebDriverError::new(
-                ErrorStatus::JavascriptError,
-                "JS evaluation raised an unknown exception",
-            )),
-            Err(WebDriverJSError::StaleElementReference) => Err(WebDriverError::new(
-                ErrorStatus::StaleElementReference,
-                "Stale element",
-            )),
-            Err(WebDriverJSError::DetachedShadowRoot) => Err(WebDriverError::new(
-                ErrorStatus::DetachedShadowRoot,
-                "Detached shadow root",
-            )),
-            Err(WebDriverJSError::UnknownType) => Err(WebDriverError::new(
-                ErrorStatus::UnsupportedOperation,
-                "Unsupported return type",
-            )),
+            Err(error) => {
+                let message = format!("{error:?}");
+                let status = match error {
+                    JavaScriptEvaluationError::DocumentNotFound => ErrorStatus::NoSuchWindow,
+                    JavaScriptEvaluationError::CompilationFailure => ErrorStatus::JavascriptError,
+                    JavaScriptEvaluationError::EvaluationFailure(Some(error_info)) => {
+                        return Err(WebDriverError::new_with_data(
+                            ErrorStatus::JavascriptError,
+                            error_info.message,
+                            None,
+                            error_info.stack,
+                        ));
+                    },
+                    JavaScriptEvaluationError::EvaluationFailure(None) => {
+                        ErrorStatus::JavascriptError
+                    },
+                    JavaScriptEvaluationError::InternalError => ErrorStatus::JavascriptError,
+                    JavaScriptEvaluationError::SerializationError(serialization_error) => {
+                        match serialization_error {
+                            JavaScriptEvaluationResultSerializationError::DetachedShadowRoot => {
+                                ErrorStatus::DetachedShadowRoot
+                            },
+                            JavaScriptEvaluationResultSerializationError::Generic => {
+                                ErrorStatus::JavascriptError
+                            },
+                            JavaScriptEvaluationResultSerializationError::StaleElementReference => {
+                                ErrorStatus::StaleElementReference
+                            },
+                            JavaScriptEvaluationResultSerializationError::UnknownType => {
+                                ErrorStatus::UnsupportedOperation
+                            },
+                        }
+                    },
+                    JavaScriptEvaluationError::WebViewNotReady => ErrorStatus::NoSuchWindow,
+                };
+                Err(WebDriverError::new(status, message))
+            },
         }
     }
 

